app/lib/task/handlers.dart

```diff
@@ -1,29 +1,108 @@
 import 'package:pub_dev/shared/exceptions.dart';
 import 'package:pub_dev/shared/handlers.dart';
+import 'package:pub_dev/shared/redis_cache.dart';
 import 'package:pub_dev/task/backend.dart';
 import 'package:shelf/shelf.dart' as shelf;
 
+const _safeMimeTypes = {
+  // Binary image files are generally safe to serve
+  'avif': 'image/avif',
+  'avi': 'video/x-msvideo',
+  'bmp': 'image/bmp',
+  'gif': 'image/gif',
+  'ico': 'image/vnd.microsoft.icon',
+  'jpeg': 'image/jpeg',
+  'jpg': 'image/jpeg',
+  'mp4': 'video/mp4',
+  'mpeg': 'video/mpeg',
+  'ogv': 'video/ogg',
+  'png': 'image/png',
+  'webm': 'video/webm',
+  'webp': 'image/webp',
+  // JSON files are generally safe to serve
+  'json': 'application/json',
+
+  // Following mimetypes can contain scripts and contents should be sanitized
+  // before we serve contents of such files.
+  //   'html: 'text/html',
+  //   'svg': 'image/svg+xml',
+  // We should always be careful about adding new mimetypes here.
+  // If it's not safe to proxy the mimetype from an untrusted source, then we
+  // should not add it here.
+};
+
+final _utf8gzip = const Utf8Codec(allowMalformed: true).fuse(gzip);
+
 Future<shelf.Response> handleDartDoc(
   shelf.Request request,
   String package,
   String version,
   String path,
 ) async {
   InvalidInputException.checkPackageName(package);
-  InvalidInputException.checkSemanticVersion(version);
+  version = InvalidInputException.checkSemanticVersion(version);
+
+  final ext = path.split('.').last;
+
+  // Handle HTML requests
+  final isHtml = ext == 'html' || ext == 'htm';
+  if (isHtml) {
+    final html = await cache.dartdocHtml(package, version, path).get(() async {
+      try {
+        final dataGz = await taskBackend.dartdocFile(package, version, path);
+        if (dataGz == null) {
+          return ''; // store empty string for missing data
+        }
+        final latestVersion = await packageBackend.getLatestVersion(package);
+        final page = DartDocPage.parse(_utf8gzip.decode(dataGz));
+        final html = page.render(DartDocPageOptions(
+          package: package,
+          version: version,
+          isLatestStable: version == latestVersion,
+          path: path,
+        ));
+        return html.toString();
+      } on FormatException {
+        // store empty string for invalid data, we treat it as a bug in
+        // the documentation generation.
+        return '';
+      }
+    });
+    // We use empty string to indicate missing file or bug in the file
+    if (html == null || html.isEmpty) {
+      return notFoundHandler(request);
+    }
+    return htmlResponse(html);
+  }
 
-  final bytes = await taskBackend.dartdocPage(package, version, path);
-  if (bytes == null) {
+  // Handle any non-HTML request
+  final mime = _safeMimeTypes[ext];
+  if (mime == null) {
+    // TODO: Communicate that this file type is not allowed!
+    // Probably we should just do this in pub_worker and write something in the log
```

Comment:
Also: will we prune the links to such file from the generated HTML?

Reply:
Probably not. If want to do so we should build it into `dartdoc`.

Afaik, no problems exist here by default, it's only if we enable tools and/or raw html in dartdoc that this likely to become an issue.